raft/Raft.tla

```diff
@@ -0,0 +1,247 @@
+LogsUpToDate(v, c) == LET lastLog == LastLogOf(v)
+                          candidateLastLog == LastLogOf(c)
+                      IN \/ candidateLastLog.term > lastLog.term
+                         \/ /\ candidateLastLog.term = lastLog.term
+                            /\ candidateLastLog.index >= lastLog.index
+
+(* Votes for any suitable cnadidate. It's the same as handling MsgRequestVote. *)
+Vote(v) ==
+    /\ \E l \in voters: /\ softState[l].role = "Candidate"
+                        /\ \/ hardState[l].term > hardState[v].term
+                           \/ /\ hardState[l].term = hardState[v].term
+                              /\ hardState[v].voteFor = None
+                        /\ LogsUpToDate(v, l)
+                        /\ hardState' = [hardState EXCEPT ![v] = [@ EXCEPT !.voteFor = l, !.term = hardState[l].term]]
+                        /\ softState' = [softState EXCEPT ![v] = [leader |-> None, role |-> "Follower"]]
+    /\ UNCHANGED <<voters, logs, history>>
+
+(* Proposes data as a log entry. *)
+AppendLog(v, data) ==
+    logs' = [logs EXCEPT ![v] = @ \union { [index |-> LastLogOf(v).index + 1, term |-> hardState[v].term, data |-> data] }]
+
+(* Becomes leader if possible. It's the same as handling RequestVoteResponse. *)
+ClaimLeadership(v) ==
+    /\ LET met == { l \in voters: /\ hardState[l].term = hardState[v].term
+                                  /\ hardState[l].voteFor = v }
+       IN Cardinality(met) * 2 > Cardinality(voters)
+    /\ softState' = [softState EXCEPT ![v] = [leader |-> v, role |-> "Leader"]]
+    /\ AppendLog(v, NoValue)
+    /\ leaderHistory' = leaderHistory \union { [leader |-> v, term |-> hardState[v].term] }
+    /\ UNCHANGED <<voters, hardState, commitHistory>>
+
+(* Checks if it's possible to commit logs. It's the same as handling MsgAppendResponse. *)
+Commit(v) ==
+    /\ LastLogOf(v).index # hardState[v].commit
+    /\ LET quorumLogs == { l \in logs[v]: /\ l.index > hardState[v].commit
+                                          /\ LET met == { o \in voters: l \in logs[o] }
+                                             IN Cardinality(met) * 2 > Cardinality(voters)}
+       IN /\ quorumLogs # {}
+          /\ LET quorumLastLog == LastLog(quorumLogs)
+             IN /\ quorumLastLog.term = hardState[v].term
+                /\ hardState' = [hardState EXCEPT ![v] = [@ EXCEPT !.commit = quorumLastLog.index]]
+                /\ commitHistory' = commitHistory \union { l \in logs[v]: /\ l.index <= quorumLastLog.index
+                                                                          /\ l.index > hardState[v].commit }
+                /\ UNCHANGED <<voters, softState, logs, leaderHistory>>
+
+StepFollower(v) == \/ CheckLeader(v)
+                   \/ FetchLog(v)
+                   \/ Vote(v)
+
+StepCandidate(v) == \/ CheckLeader(v)
+                    \/ Vote(v)
+                    \/ ClaimLeadership(v)
+
+StepLeader(v) == \/ CheckLeader(v)
+                 \/ Commit(v)
```

Comment:
Does this mean we can push commits even it is not a leader?

Reply:
Yes, it works like message delay. Even there is a new leader elected, delay `MessageAppendResponse` can still make old leader advance commit index.

Reply:
But message delay should rarely happen in the real world?

Reply:
Correctness should not rely on *rarely happen*.

Reply:
got